### PR TITLE
Add: PKCE flow for android native

### PR DESCRIPTION
**Pull request by nishihere19**
Signed-off-by: nishihere19 <nishihere19@gmail.com>

Fixes #

## Short description of the Issue.
PKCE for android apps

## Proposed Changes

  - Added the PKCE flow for android apps
  - The query params are optional and authorization works normally for the web clients.

### CODE REVIEW

File: src/routes/oauth.ts [hunk 1]
```diff
@@ -13,13 +13,17 @@ import oauth from '../oauth/index';
...
+// redirect_uri            |
+// response_type           |
+// grant_type              | -> query params
+// state                   |
+// scope                   |
+// nonce                   |
+
+// code_challenge          |
+// code_challenge_method   | ->pkce
+// code_challenge_verifier |
```
Comment (siva2204) [inline src/routes/oauth.ts:26]: this is not needed ig, not used any where too,can be removed

File: src/controllers/oauthContoller.ts [hunk 2]
```diff
@@ -32,6 +36,24 @@ export const handleAuthorize = async (req: Request, res: Response) => {
     //check if client is authorized by the user
     const isAuthorized = await isAuthorizedApp(code.client.primaryKey, user.id);
 
+    //check if pkce flow exists(android clients)
```
Comment (siva2204) [inline src/controllers/oauthContoller.ts:39]: this check should be done in token endpoint ig.

**Review by siva2204 — CHANGES_REQUESTED** [review]
PKCE is to safe the native apps from authorization **code interception**.

so the code (code_verifier) should be verified in token route i.e while client exchange authorization code for a token.

- [flow digram](https://datatracker.ietf.org/doc/html/rfc7636#section-1.1)
- [complete-protocol](https://datatracker.ietf.org/doc/html/rfc7636#section-1.1)

we should save code_challenge,code_challenge_method linked to the authorization_code in server side(preferably db) and these saved code_challenge should be used to authorize when client asks for token in exchange of authorization code.

**Review by siva2204 — CHANGES_REQUESTED** [review]
we should think of way to make all the native apps to follow PKCE.

all we need is a proper documentation and examples for PKCE as wells as OAuth.

TODO:
- better documentation
- examples for web and native

**Comment by nishihere19** [comment]
Sure, will implement.

File: src/controllers/oauthContoller.ts [hunk 3]
```diff
@@ -131,6 +132,44 @@ export const getClaims = async (
...
+    if (codeData?.codeChallenge) {
+      if (!req.body.code_verifier) {
+        return res.status(400).json({ message: 'code_verifier is missing' });
+      }
+      const isVerified = await verifyCodeChallenge(
+        code,
+        req.body.code_verifier
+      );
+      if (!isVerified) {
+        return res.status(400).json({ message: 'Internal server error.' });
```
Comment (siva2204) [inline src/controllers/oauthContoller.ts:164]: return proper error if verification fails, check statandard.

as i can see `verifyCodeChallenge` return bool and throws error.
return `internal server error` if verifyCodeCahellenge throws an error.

File: src/controllers/oauthContoller.ts [hunk 4]
```diff
@@ -131,6 +132,44 @@ export const getClaims = async (
...
+  try {
+    const codeData = await prisma.code.findUnique({
+      where: {
+        code: code
+      },
+      select: {
+        codeChallenge: true
+      }
+    });
+    if (codeData?.codeChallenge) {
```
Comment (siva2204) [inline src/controllers/oauthContoller.ts:155]: if there is a codeChallenge associated with code(authorization code), and code_verifier is missing in the request return proper error in response.

File: src/utils/oauth.ts [hunk 5]
```diff
@@ -76,8 +80,24 @@ export const saveStateAndNonce = async (
         state: state
       }
     });
+    if (codeChallenge) {
```
Comment (siva2204) [inline src/utils/oauth.ts:83]: its better to create a new utility func to save data regarding to pkce.

**Review by siva2204 — CHANGES_REQUESTED** [review]

**Comment by siva2204** [comment]
others looks fine, can merge after these changes.

**Pull request merged**